### ISSUE TITLE
(we need a) Better way to sync CI tests with `toolkit check pr`

### ISSUE THREAD

**Issue by bobhy** [nushell/nushell#10729]
# Related Problem

Currently, (and often in the history of the project), the tests run by CI are subtly different from what was last captured in `toolkit check pr`, so dev gets CI failures that "should" have been detected by the local run of `toolkit`.  It would be great if there could be better coordination between these tools. But CI should be the source of truth regarding the the details of tests run by `check pr`.  

Here's a recent conversation about it
https://github.com/nushell/nushell/issues/10694#issuecomment-1763268665_

and some information about where CI tests are configured:
https://github.com/nushell/nushell/pull/10694#issuecomment-1763379906

# Describe the solution you'd like

Running `toolkit check pr` should always detect the same set of issues that CI would detect for the cells in its configuration matrix that can reasonably be run in developer's environment.

I think that means running tests with the same cargo switches and running them against a "reasonable" set of feature configs on dev's brand of OS (ignoring version/release).  I don't think it should require docker or other VM test runner capability on the dev machine -- that is the domain of CI on github.

# Alternatives you've considered

1. Wrap each test that CI runs in a (sh) shell script that is checked into the project along with `ci.yaml` so `toolkit.nu` can easily invoke the same command line.
2. Have `toolkit.nu` parse `ci.yaml` and predict the command line that github actions would generate.  (Less intrusion into ci, but less certain of the desired result.  Ci.yaml is a proprietary API, though not undocumented.)

**Comment by bobhy** [comment]
A bit more googling, and I found this project: https://github.com/nektos/act.
Anybody have any experience with this?

It looks like it recreates the whole github actions environment in Docker containers, and would allow local CI runs of the whole test matrix (if you have windows and macOS licenses).

But this is only for those with truly heavyweight dev machines or a private server, not for all nushell devs.

**Comment by amtoine** [comment]
i completely agree with you @bobhy :+1: 

btw, this is not only `toolkit check pr` but mostly `toolkit fmt`, `toolkit clippy`, `toolkit test` and `toolkit test stdlib`, 'cause `toolkit check pr` only calls these ones :open_mouth: 

maybe these commands could try to parse the `ci.yml` file? :thinking:

**Comment by bobhy** [comment]
Agreed, I see `check pr` invokes the lower-level functions.  Note that
`fmt` when invoked directly will *correct* the issues it finds (`--check`
defaults to false), while, when invoked from `check pr`, it simply
*reports* the problems (with `--check` specified); reporting issues is just
the behavior you want before pushing the PR.

I would not want to put the logic in `toolkit.nu` to parse the `ci.yaml`
file, I think it's much more robust and maintainable for CI to publish an
API (e.g shell scripts wrapping each test) that can be invoked outside the
CI environment.

**Comment by bobhy** [comment]
That might work, though the `nu` executable that CI runs should not be the
one built from current sources.  It would be a problem if a bug in the
current source tree caused unrelated failures in CI...
